Added new pages for optimization, singletons, and pooling (#54)
Chia-Network/chialisp-web: docs/puzzles/pooling.md

---
id: pooling
title: Pooling
---

> The way that Chia Network does pooling is unlike any blockchain that has come before it. Pool operators actually rely on an on-chain smart contract to verify that they will be able to directly claim any potential pool rewards that farmers create.  This allows pools to trust farmers enough to pay them out, while still keeping the power of making the blocks in the hands of the farmer.  This means that the decentralization of the network remains the same, even as the rewards get concentrated to the pool!

Reviewer: "...any blockchain..." This seems like a difficult claim to prove... maybe we should undersell a bit.

Reviewer: There are over 2500 coins on coinmarketcap.com and I doubt that ANYONE is familiar with all of them.

Author: If someone already did this, that would make me sad.  But okay, I'll change it to "unlike many blockchains that have come before it"

In this section, we're going to break down how all of this works in Chialisp.  This section assumes you have already read the section about [singletons](https://chialisp.com/docs/puzzles/singletons) (or at least understand how they work) as that is the outer contract that wraps the pooling puzzles.

## Design Requirements

There are a few requirements that were set for how the pooling protocol would work on the Chia Network. Let's go over them now:

> * **The farmer still farms the blocks.** This is incredibly important for network decentralization. If this is not true, then the bigger a pool gets, the closer it is to gaining 51% of the network resources.  We still want the farmers to create the blocks, but we need some way to assure the pool that the farmer will give them the reward when they do.  We do this by creating plots that farm directly to a specific puzzle hash, and ensuring that puzzle hash is something that pool can claim.

Reviewer 2: It feels like there's a missing "the" in front of "pool" here

> * **The farmer still farms the blocks.** This is incredibly important for network decentralization. If this is not true, then the bigger a pool gets, the closer it is to gaining 51% of the network resources.  We still want the farmers to create the blocks, but we need some way to assure the pool that the farmer will give them the reward when they do.  We do this by creating plots that farm directly to a specific puzzle hash, and ensuring that puzzle hash is something that pool can claim.

Reviewer 3: `The farmer still farms the blocks` ->
The farmer farms the blocks, not the pool.

> * **The farmer must still be able to change pools.** Initially, this seems to conflict with the first requirement.  Plots can be made to send rewards directly to a puzzle hash, but you cannot change that puzzle hash once they are plotted.  If you want to switch pools, you'll have to remake all of your plots! We solve this by having our plots create payments that a specific singleton (also called a "plot nft") can claim.  Then, we can lend partial control of that singleton to a pool, but still retain the ability to reclaim our singleton and lend it instead to a different pool.  Our plots will remain effective as long as we retain control of the singleton.

Reviewer 2: The term "singleton" isn't explained here. Maybe this term is established earlier, but if it isn't, it might be worth just describing it as a "smart contract"

Author: The top of the page says that it assumes knowledge of singletons if that counts.

    )
  )

  (defun-inline travel_to_waitingroom (OWNER_PUBKEY WAITINGROOM_PUZHASH my_amount extra_data)
    (list (list AGG_SIG_ME OWNER_PUBKEY (sha256tree extra_data))
          (list CREATE_COIN WAITINGROOM_PUZHASH my_amount)
    )
  )

  ; main

  (if pool_reward_height
    (absorb_pool_reward POOL_PUZZLE_HASH
                        (my_inner_puzzle_hash_truth Truths)
                        (my_amount_truth Truths)
                        p1
                        (calculate_pool_reward pool_reward_height P2_SINGLETON_PUZZLE_HASH POOL_REWARD_PREFIX p1)
    )
    (travel_to_waitingroom OWNER_PUBKEY WAITINGROOM_PUZHASH (my_amount_truth Truths) p1)
  )
)
```

As always, let's begin with the arguments:

`POOL_PUZZLE_HASH` is the address to which we want to pay the pool when we win a block reward.  The pool uses this to verify that the singleton has been lent to them.

`P2_SINGLETON_PUZZLE_HASH` is the puzzle hash that payments to this singleton will have.  This is the hash that farmers make their plots to.

`OWNER_PUBKEY` is the public key that will sign the decision to leave the pool.  This is likely owned by the farmer.

> `POOL_REWARD_PREFIX` is a bit of a unique argument.  On the chia mainnet, this will always be `ccd5bb71183532bff220ba46c268991a00000000000000000000000000000000`.  It is the beginning of the `parent_coin_info` that is assigned to the pool rewards since they do not have a parent.  That first half is followed by 32 zeroes to make it the length of a standard hash.  We'll talk more about why this is needed when we go over the segment of code that uses it.

Reviewer 3: ```suggestion
`POOL_REWARD_PREFIX` is a bit of a unique argument.  On the chia mainnet, this will always be `ccd5bb71183532bff220ba46c268991a00000000000000000000000000000000`.  It is the beginning of the `parent_coin_id` of pool reward coins. Because those coins do not have a parent, their ID is derived from a combination of the network ID in the left half of the ID, and an incrementing identifier in the right.  That first half is followed by 32 zeroes to make it the length of a standard hash.  We'll talk more about why this is needed when we go over the segment of code that uses it.
```

`WAITINGROOM_PUZHASH` is the puzzle hash of the waiting room puzzle that we will go to when we attempt to leave the pool.  We will go over that puzzle later, but what's important to know is that we commit to the puzzle we leave to before we join the pool.  This is so the pool can have certain assurances about how long it will take you to leave.

`Truths` is the data structure that is forcefully added to this puzzle as part of the singleton top layer.  It contains some data we will use in the puzzle which we will access by using some functions from the `singleton_truths.clib` library.

There are two spend cases for this puzzle.  The **absorb case** will most likely be triggered by the pool and is how they claim the rewards that the farmers receive.  The **escape case** will be initiated by the farmer to begin the process of leaving the pool by heading to the waiting room.  The following two arguments change depending on the case we are executing:

`p1` is named the way it is because it is going to be different depending on the spend type we are using.  It is either:

>  * The amount of the pool reward that is being claimed (1750000000000 in the first three years)

Reviewer 3: ```suggestion
 * The amount of the pool reward that is being claimed (1750000000000 during the first three years)
```

```chialisp
(if pool_reward_height
  (absorb_pool_reward POOL_PUZZLE_HASH
                      (my_inner_puzzle_hash_truth Truths)
                      (my_amount_truth Truths)
                      p1
                      (calculate_pool_reward pool_reward_height P2_SINGLETON_PUZZLE_HASH POOL_REWARD_PREFIX p1)
  )
  (travel_to_waitingroom OWNER_PUBKEY WAITINGROOM_PUZHASH (my_amount_truth Truths) p1)
)
```

The main control flow here is whether or not the height is passed in as `()`. This is how we determine the spend type to execute. After that we just head to the appropriate function that generates the conditions we will need for the type of spend we are executing. You'll notice that we are extracting the relevant data from the `Truths` object before we pass them to the inner functions.

We are also calculating one additional piece of information.  Let's look at it now:

```chialisp
(defun-inline calculate_pool_reward (pool_reward_height P2_SINGLETON_PUZZLE_HASH POOL_REWARD_PREFIX pool_reward_amount)
  (sha256 (logior POOL_REWARD_PREFIX (logand (- (lsh (q . 1) (q . 128)) (q . 1)) pool_reward_height)) P2_SINGLETON_PUZZLE_HASH pool_reward_amount)
)
```

> This line may look complex, but it's doing something fairly simplistic.  It is calculating the coin ID of the reward coin it is claiming by manually calculating the parent info from the `POOL_REWARD_PREFIX` and `pool_reward_height`.

Reviewer 2: This "parent info" isn't referring to parent coin is, right? (Since that's being renamed right now)

> Why do we have to do this? The manual calculation is due to the fact that this singleton may have other payments made to it.  Right now, since we are lending our singleton to the pool, we don't want the singleton to be able to claim those rewards. We specifically only want the pool to be able to claim pool rewards that are generated from farming.  Since we know these rewards have a `parent_coin_info` that is a [special format](https://chialisp.com/docs/coin_lifecycle#farming-rewards), we can manually calculate it to ensure that the pool can't lie to us and pass in the ID of a non-reward coin.

Reviewer 2: This "parent_coin_info" looks like it should be "parent_coin_id" at least

Let's take a look at this section here:

```chialisp
(logior POOL_REWARD_PREFIX (logand (- (lsh (q . 1) (q . 128)) (q . 1)) pool_reward_height))
```

> `(- (lsh (q . 1) (q . 128)) (q . 1))` is simply a way of generating a sequence of 32 `f`s. We then use `logand` on that with the pool reward height, which in an honest scenario should do absolutely nothing to it (more on this in a minute).  Finally, we use `logior` to combine the two values into one string.  Let's say we have a height of `abcdef`.  Our final product will be `ccd5bb71183532bff220ba46c268991a00000000000000000000000000abcdef`.

Reviewer 3: ```suggestion
`(- (lsh (q . 1) (q . 128)) (q . 1))` is simply a way of generating a sequence of 32 `f`s. We then use `logand` on that with the pool reward height, which in an honest scenario should leave it unchanged.  Finally, we use `logior` to combine the two values into one string.  Let's say we have a height of `abcdef`.  Our final product will be `ccd5bb71183532bff220ba46c268991a00000000000000000000000000abcdef`.
```

> Why do we need this extra `logand` with the string of `f`s?  It's to prevent a relatively obscure, but possible attack.  Remember that an attacker can pass whatever they want in through the solution.  For example, a 32 byte hex string.  If they passed through the right hex string, they could completely control the output of our calculation *except* for the bits that happen to be set to 1 in the genesis challenge half of the `POOL_REWARD_PREFIX` (62/128 of them).  The idea is that a pool could grind out a coin whose parent ID has all of those bits set, create the coin, and then use the singleton to claim it.  Why claim a coin that you already had control of? This will become more apparent in the waiting room puzzle, but they could hypothetically "freeze" the singleton in their pool if they were able to reset the puzzle.  It's not as important in the pool member puzzle, but it is still probably best to prevent the pool from spending the singleton in an unwanted way.  The `logand` ensures that only bits on the right can change, every bit on left gets set to 0 before it is evaluated with the `logior`.

Reviewer 3: ```suggestion
Why do we need this extra `logand` with the string of `f`s?  It's to prevent a relatively obscure, but possible attack.  Remember that an attacker can pass whatever they want in through the solution.  For example, a 32 byte hex string.  If they passed through the right hex string, they could completely control the output of our calculation *except* for the bits that happen to be set to 1 in the genesis challenge half of the `POOL_REWARD_PREFIX` (62/128 of them).  The idea is that a pool could grind out a coin whose parent ID has all of those bits set, create the coin, and then use the singleton to claim it.  Why claim a coin that you already had control of? This will become more apparent in the waiting room puzzle, but they could hypothetically "freeze" the singleton in their pool if they were able to reset the puzzle.  It's not as important in the pool member puzzle, but it is still probably best to prevent the pool from spending the singleton in an unintended way.  The `logand` ensures that only bits on the right can change, every bit on left gets set to 0 before it is evaluated with the `logior`.
```

          (list AGG_SIG_ME OWNER_PUBKEY (sha256tree (list new_puzzle_hash extra_data)))
    )
  )

  ; main

  (if spend_type
    (travel_spend RELATIVE_LOCK_HEIGHT p2 (my_amount_truth Truths) p1)
    (absorb_pool_reward POOL_PUZZLE_HASH
                        (my_inner_puzzle_hash_truth Truths)
                        (my_amount_truth Truths)
                        p1
                        (calculate_pool_reward p2 P2_SINGLETON_PUZZLE_HASH POOL_REWARD_PREFIX p1)
    )
  )
)
```

You may notice that is looks nearly identical to the one above.  Instead of breaking this down piece by piece, we're just going to focus on the differences.  First, the parameters:

```chialisp
(
  POOL_PUZZLE_HASH
  P2_SINGLETON_PUZZLE_HASH
  OWNER_PUBKEY
  POOL_REWARD_PREFIX
  RELATIVE_LOCK_HEIGHT
  Truths
  spend_type
  p1
  p2
)
```

We still have `POOL_PUZZLE_HASH`, `P2_SINGLETON_PUZZLE_HASH`, `OWNER_PUBKEY`, and `POOL_REWARD_PREFIX`.  However, now we also have a new curried in parameter called `RELATIVE_LOCK_HEIGHT`. This indicates the amount of time after entering this waiting room that we have to wait before we can spend away to something else.

> Note that relative lock heights are calculated from the time of the *current coin's creation*.  If the coin is spent as part of an absorb, this lock height resets.  Theoretically, with a large enough lock height and a big enough farmer who wins frequently, this singleton can be "frozen" until the farmer is lucky enough not to win a block within the specified timeframe.

Reviewer: Maybe drop the word "current", think it's actually more clear without it.
